File: scripts/causal-position-pairs.test.mjs
/** D169 loaded qualification. Timing loops use fake clocks and stub factories. */
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import { syncBuiltinESMExports } from "node:module";
import os from "node:os";
import path from "node:path";
import { performance } from "node:perf_hooks";
import timers from "node:timers/promises";
import { fileURLToPath, pathToFileURL } from "node:url";
import { check } from "./check-causal-position-source.mjs";
import { derive } from "./derive-causal-position-driver.mjs";

const self = fileURLToPath(import.meta.url);
const material = process.argv[2];
assert.ok(material, "explicit offline material directory required");
const original = fs.readFileSync(path.join(material, "worker.mjs"), "utf8");
const fixed = {
	revision: "spending-preset-performance-v1",
	coldRows: 12,
	steadyRows: 60,
	recoveryRows: 12,
	warmup: 100,
	measured: 300,
	orders: [
		["candidate", "reference"],
		["reference", "candidate"],
		["candidate", "reference"],
	],
	coldLimit: 1.2,
	steadyLimit: 1.1,
	recoveryCycles: 20,
	stopAfterFailedRow: true,
	childTimeoutMs: 900000,
	totalTimeoutMs: 7200000,
	freshBasis: "distinct evaluation identities absent before the whole wave",
	doubleData:
		"two exact copies of the same arrival frame in one source.down; second copy is intra-wave replay",
	memory:
		"raw process heap/RSS before and after action; GC may make deltas negative, not retained-size proof",
};
if (process.argv[3] === "--stub") {
	const [driver, kind, orientation, mutation] = process.argv.slice(4);
	const n = await import(pathToFileURL(driver));
	const orders =
		orientation === "P"
			? [["plain"], ["plain"], ["plain"]]
			: fixed.orders.map((row) => (orientation === "U" ? row : [...row].reverse()));
	const config = {
		row: { id: "cold-P2-summary", group: "cold", profile: "P2", mode: "summary" },
		kind,
		orientation,
		control: kind === "control",
		output: "/stub",
		scenarioPath: "scenario",
	};
	if (mutation === "row") config.row.group = "recovery";
	if (mutation === "kind") config.kind = "other";
	if (mutation === "orientation") config.orientation = "X";
	if (mutation === "plain-order") config.orientation = "P";
	if (mutation === "control") config.control = !config.control;
	const freeze = (x) => {
		if (x && typeof x === "object") {
			Object.values(x).forEach(freeze);
			Object.freeze(x);
		}
		return x;
	};
	freeze(fixed);
	let now = 0,
		serial = 0;
	const events = [],
		samples = [],
		outputs = {};
	const modules = [0, 1].map((id) => ({
		RECIPE: fixed,
		preflight() {
			events.push(["preflight", id]);
			return { passed: true };
		},
		measurementArm(arm, mode) {
			assert.equal(mode, "summary");
			const instance = ++serial;
			events.push(["factory", id, arm, instance]);
			if (mutation === "factory-failure") throw Error("factory");
			return {
				cleanup() {
					events.push(["cleanup", instance]);
					if (mutation === "dual-failure" || mutation === "extra-cleanup") throw Error("cleanup");
				},
			};
		},
	}));
	let configPath = "config",
		entryPath;
	const loads = [];
	if (mutation === "entry") {
		const parent = path.dirname(driver),
			job = path.join(parent, "entry-job");
		fs.mkdirSync(job, { recursive: true });
		config.output = job;
		config.scenarioPath = path.join(parent, "P2-inputs.json");
		configPath = path.join(job, "config.json");
		entryPath = path.join(job, "entry.mjs");
		globalThis.positionStubModules = modules;
		for (const [id, name] of [
			[0, "worker.mjs"],
			[1, "worker-copy.mjs"],
		]) {
			fs.writeFileSync(
				path.join(parent, name),
				`const m = globalThis.positionStubModules[${id}]; export const RECIPE=m.RECIPE; export const preflight=(...args)=>m.preflight(...args); export const measurementArm=(...args)=>m.measurementArm(...args);`,
			);
		}
		const generated = spawnSync(
			"python3",
			[
				"-c",
				"import importlib.util,sys; from pathlib import Path; s=importlib.util.spec_from_file_location('r',sys.argv[1]); r=importlib.util.module_from_spec(s); s.loader.exec_module(r); print(r.entry_source(Path(sys.argv[2]),dict(kind=sys.argv[3],orientation=sys.argv[4])),end='')",
				path.resolve("scripts/causal-position-pairs.py"),
				job,
				kind,
				orientation,
			],
			{ encoding: "utf8", timeout: 15000 },
		);
		assert.equal(generated.status, 0, generated.stderr);
		fs.writeFileSync(entryPath, generated.stdout);
	}
	const realRead = fs.readFileSync;
	fs.readFileSync = (p, ...args) =>
		p === configPath
			? JSON.stringify(config)
			: p === config.scenarioPath
				? "{}"
				: realRead(p, ...args);
	fs.writeFileSync = (p, s) => {
		outputs[p] = JSON.parse(s);
	};
	fs.appendFileSync = (_p, s) => {
		if (_p.endsWith("entry-events.jsonl")) {
			loads.push(JSON.parse(s).module);
			return;
		}
		events.push(["record"]);
		if (mutation === "dual-failure") throw Error("record");
		samples.push(JSON.parse(s));
	};
	timers.setImmediate = async () => {
		events.push(["tick"]);
	};
	Object.defineProperty(performance, "now", {
		configurable: true,
		value: () => {
			events.push(["clock"]);
			return ++now;
		},
	});
	process.memoryUsage = () => {
		events.push(["memory"]);
		return {};
	};
	syncBuiltinESMExports();
	let failure;
	try {
		if (entryPath) await import(pathToFileURL(entryPath));
		else await n.runRow(configPath, mutation === "namespace" ? [modules[0], modules[0]] : modules);
	} catch (e) {
		failure = e;
	}
	if (["row", "kind", "orientation", "plain-order", "control", "namespace"].includes(mutation)) {
		assert.ok(failure);
		assert.deepEqual(events, []);
	} else if (mutation === "dual-failure") {
		assert.ok(failure instanceof AggregateError);
		assert.deepEqual(
			failure.errors.map((e) => e.message),
			["record", "cleanup"],
		);
		assert.equal(serial, 1);
	} else if (mutation === "extra-cleanup") {
		assert.ok(failure);
		assert.equal(serial, 1);
		assert.deepEqual(events.slice(-2), [
			["factory", 1, "reference", 1],
			["cleanup", 1],
		]);
	} else if (mutation === "factory-failure") {
		assert.equal(failure?.message, "factory");
		assert.equal(serial, 1);
	} else {
		assert.equal(failure, undefined);
		assert.deepEqual(
			events.slice(0, kind === "plain" ? 1 : 2),
			kind === "plain"
				? [["preflight", 0]]
				: [
						["preflight", 0],
						["preflight", 1],
					],
		);
		let cursor = events.findIndex((e) => e[0] === "tick"),
			id = 0;
		for (const arms of orders)
			for (const arm of arms)
				for (let i = 0; i < 400; i++) {
					const expected = [["tick"], ["memory"], ["clock"]];
					const slot = arm === "candidate" ? 1 : 0;
					const selected =
						arm === "plain"
							? "plain"
							: arm === "candidate" && kind === "main"
								? "candidate"
								: "reference";
					if (kind === "mutation" && arm === "candidate") {
						expected.push(["factory", slot, selected, ++id], ["cleanup", id]);
					}
					expected.push(
						["factory", slot, selected, ++id],
						["clock"],
						["memory"],
						["record"],
						["cleanup", id],
					);
					assert.deepEqual(events.slice(cursor, cursor + expected.length), expected);
					cursor += expected.length;
				}
		assert.equal(cursor, events.length);
		assert.equal(serial, kind === "mutation" ? 3600 : kind === "plain" ? 1200 : 2400);
		assert.equal(samples.length, kind === "plain" ? 1200 : 2400);
		assert.ok(samples.every((s) => s.ms === 1 && s.constructionMs === 0 && s.preparationMs === 0));
		assert.deepEqual(outputs[`${config.output}/completion.json`], {
			completed: true,
			samples: samples.length,
		});
		assert.deepEqual(outputs[`${config.output}/worker.json`].recipe, { ...fixed, orders });
		if (entryPath) {
			assert.deepEqual(loads, ["N", "M0", "M1"]);
			assert.deepEqual(outputs[`${config.output}/recipe-after.json`].recipe, fixed);
		}
	}
	console.log(
		JSON.stringify({
			kind,
			orientation,
			mutation,
			passed: true,
			instances: serial,
			samples: samples.length,
			fakeClock: true,
		}),
	);
} else {
	const tmp = fs.mkdtempSync(path.join(os.tmpdir(), "causal-position-qualification-"));
	try {
		const { output, manifest } = derive(original);
		check(original, output);
		const driver = path.join(tmp, "position.mjs");
		fs.writeFileSync(driver, output);
		let negatives = 0;
		const mutations = [
			["const arms = RECIPE.orders[batch2];", "const arms = [...RECIPE.orders[batch2]].reverse();"],
			['const slot = arm === "candidate" ? 1 : 0;', 'const slot = arm === "candidate" ? 0 : 1;'],
			['kind === "mutation" && arm === "candidate"', "false"],
			["cleanupAll([extra]);", "void extra;"],
			[
				"const extra = modules[slot].measurementArm(selected,mode);",
				"const extra = {cleanup(){}};",
			],
			[
				"return modules[slot].measurementArm(selected,mode);",
				"return modules[0].measurementArm(selected,mode);",
			],
			['const selected = arm === "plain"', 'const selected = arm === "reference"'],
			["assert.deepEqual(modules[0].RECIPE,originalRecipe);", ""],
			["assert.deepEqual(modules[1].RECIPE,originalRecipe);", ""],
			[
				"const RECIPE = {...originalRecipe,orders:orders[config.orientation]};",
				"const RECIPE = modules[0].RECIPE;",
			],
			[
				"const checked3 = modules[0].preflight(row,scenario);",
				"const checked3 = modules[1].preflight(row,scenario);",
			],
			['config.kind !== "plain"', "false"],
			["const start = performance2.now();", "const start = 0;"],
			["const end = performance2.now();", "const end = start;"],
			["ms: end - start", "ms: 2"],
			["if (run !== shared) cleanupAll([run], sampleFailure);", ""],
			["appendFileSync(samplesPath,", "writeFileSync(samplesPath,"],
			["batch2 < RECIPE.orders.length", "batch2 < 2"],
			["index < total", "index < total - 1"],
			["await setImmediate();", ""],
		];
		for (const [from, to] of mutations) {
			assert.ok(output.includes(from), from);
			assert.throws(() => check(original, output.replace(from, to)));
			negatives++;
		}
		assert.throws(() => check(original, output + 'console.log("extra");'));
		negatives++;
		const cases = [
			["control", "U", "none"],
			["control", "V", "none"],
			["main", "U", "none"],
			["main", "V", "none"],
			["mutation", "U", "none"],
			["mutation", "V", "none"],
			["plain", "P", "none"],
			...[
				"row",
				"kind",
				"orientation",
				"plain-order",
				"control",
				"namespace",
				"dual-failure",
				"factory-failure",
			].map((m) => ["control", "U", m]),
			["mutation", "U", "extra-cleanup"],
		];
		cases.push(
			...[
				["control", "U"],
				["control", "V"],
				["main", "U"],
				["main", "V"],
				["mutation", "U"],
				["mutation", "V"],
				["plain", "P"],
			].map(([kind, orientation]) => [kind, orientation, "entry"]),
		);
		const loaded = [];
		for (const args of cases) {
			const result = spawnSync(process.execPath, [self, material, "--stub", driver, ...args], {
				encoding: "utf8",
				timeout: 15000,
			});
			assert.equal(result.status, 0, result.stderr);
			loaded.push(JSON.parse(result.stdout.trim().split("\n").at(-1)));
		}
		const runtimeFaults = [
			mutations[1],
			mutations[2],
			mutations[3],
			mutations[4],
			mutations[5],
			mutations[14],
			mutations[17],
		];
		for (let i = 0; i < runtimeFaults.length; i++) {
			const [from, to] = runtimeFaults[i],
				file = path.join(tmp, `mutant-${i}.mjs`);
			fs.writeFileSync(file, output.replace(from, to));
			const run = spawnSync(
				process.execPath,
				[self, material, "--stub", file, "mutation", "U", "none"],
				{ encoding: "utf8", timeout: 15000 },
			);
			assert.notEqual(run.status, 0, `loaded mutant ${i} escaped`);
		}
		const moved = output
			.replace('kind === "mutation" && arm === "candidate"', "false")
			.replace(
				"const memoryBefore = process.memoryUsage();",
				'if (config.kind === "mutation" && arm === "candidate") { const extra=modules[1].measurementArm("reference",row.mode); cleanupAll([extra]); }\nconst memoryBefore = process.memoryUsage();',
			);
		const swallowed = output.replace(
			"cleanupAll([extra]);",
			"try { cleanupAll([extra]); } catch {}",
		);
		for (const [name, source, fault] of [
			["outside-timer", moved, "none"],
			["swallowed-cleanup", swallowed, "extra-cleanup"],
		]) {
			assert.notEqual(source, output);
			assert.throws(() => check(original, source));
			negatives++;
			const file = path.join(tmp, `${name}.mjs`);
			fs.writeFileSync(file, source);
			const run = spawnSync(
				process.execPath,
				[self, material, "--stub", file, "mutation", "U", fault],
				{ encoding: "utf8", timeout: 15000 },
			);
			assert.notEqual(run.status, 0, `${name} loaded source escaped`);
		}
		// Append read-only qualification exports to immutable bytes; no mutation of factory/oracle bodies.
		const appended =
			"\nexport {graphSnapshot as qualificationSnapshot,verifyBusiness as qualificationOracle,sorted as qualificationSorted};\n";
		const modules = [];
		for (let i = 0; i < 2; i++) {
			const file = path.join(tmp, `real-${i}.mjs`);
			fs.writeFileSync(file, original + appended);
			assert.equal(fs.readFileSync(file, "utf8").slice(0, original.length), original);
			modules.push(await import(pathToFileURL(file)));
		}
		const n = await import(pathToFileURL(driver));
		const scenario = JSON.parse(fs.readFileSync(path.join(material, "P2-inputs.json"), "utf8"));
		const real = [];
		for (const kind of ["control", "main", "mutation"]) {
			let created = 0,
				cleaned = 0;
			const live = new Set();
			const wrapped = modules.map((m) => ({
				RECIPE: m.RECIPE,
				measurementArm(arm, mode) {
					const run = m.measurementArm(arm, mode);
					created++;
					live.add(run);
					return {
						...run,
						cleanup() {
							assert.ok(live.delete(run), "one cleanup per instance");
							run.cleanup();
							cleaned++;
							assert.equal(run.graph?.describe().nodes.length ?? 0, 0, "released graph topology");
						},
					};
				},
			}));
			const make = n.makeFactory(wrapped, kind);
			let a, b, p;
			try {
				a = make("candidate", "summary");
				assert.equal(created, kind === "mutation" ? 2 : 1);
				assert.equal(cleaned, kind === "mutation" ? 1 : 0);
				assert.equal(live.size, 1);
				b = make("reference", "summary");
				p = make("plain", "summary");
				for (const step of scenario.steps) {
					for (const run of [a, b, p]) run.send(step);
					const sa = modules[0].qualificationSnapshot(a);
					assert.deepEqual(modules[0].qualificationSnapshot(b), sa);
					assert.deepEqual(p.plain.snapshot(), sa.effects);
					assert.deepEqual(modules[0].qualificationSorted(p.plain.evidenceSnapshot()), sa.evidence);
					assert.deepEqual(
						modules[0].qualificationSorted(p.plain.obligationSnapshot()),
						sa.obligations,
					);
				}
				const assessment = a.latest.assessment;
				for (const e of scenario.evaluations) {
					const observed = assessment.rows.find(
						(r) => r.evaluation.evaluationRef === e.evaluationRef,
					);
					assert.ok(observed && modules[0].qualificationOracle(e, observed.value));
				}
				for (const run of [a, b]) {
					const before = modules[0].qualificationSnapshot(run);
					run.disconnect();
					assert.deepEqual(modules[0].qualificationSnapshot(run), before);
					run.connect();
					assert.deepEqual(modules[0].qualificationSnapshot(run).obligations, before.obligations);
				}
			} finally {
				for (const run of [a, b, p]) run?.cleanup();
			}
			assert.equal(live.size, 0);
			assert.equal(created, cleaned);
			real.push({ kind, created, cleaned, oraclePassed: true });
		}
		console.log(
			JSON.stringify({
				passed: true,
				manifest,
				staticNegatives: negatives,
				loadedCases: loaded,
				loadedSourceMutants: runtimeFaults.length + 2,
				realAdapterQualification: real,
				consumerPerformanceSamples: 0,
			}),
		);
	} finally {
		fs.rmSync(tmp, { recursive: true, force: true });
	}
}

File: scripts/causal-position-pairs.test.py
"""D169 offline test vectors and failure replay. Synthetic clocks are not measurements."""
import copy
import importlib.util
import json
import math
from pathlib import Path
import shutil
import subprocess
import tarfile
import tempfile
import unittest
from unittest.mock import patch

ROOT=Path(__file__).resolve().parents[1]
def import_tool(name):
    spec=importlib.util.spec_from_file_location(name,ROOT/'scripts'/f'{name}.py')
    x=importlib.util.module_from_spec(spec);spec.loader.exec_module(x);return x
r=import_tool('causal-position-pairs');v=import_tool('verify-causal-position-pairs')
ARCHIVE=ROOT/'archive/evals/causal-performance-repetition-v2'
NEGATIVES=0

def put(p,x):p.write_text(json.dumps(x,allow_nan=False)+'\n')
def reindex(root):put(root/'artifact-index.json',dict(files={str(p.relative_to(root)):v.sha(p) for p in root.rglob('*') if p.is_file() and p!=root/'artifact-index.json'}))
def samples(cell,z_cost=100):
    output=[];time=0
    for b in range(3):
        for s,arm in enumerate(r.ORDERS[cell['orientation']][b]):
            base=(100 if s==0 else 80)*(b+1)
            cost=2*base if cell['kind']=='mutation' and arm=='candidate' else base
            if cell['panel']=='Z' and cell['category']=='main' and arm=='candidate':cost=base*z_cost/100
            for i in range(400):
                start=time;end=start+cost;time=end+1
                output.append(dict(batch=b,arm=arm,index=i,phase='warmup' if i<100 else 'measured',segment=f'cold-P2-summary/{b}/{arm}/{i}',start=start,end=end,ms=end-start,constructionMs=0,preparationMs=0))
    return output

def fixture(root,count=80,z_cost=100):
    root.mkdir();(root/'source').mkdir();(root/'qualification').mkdir()
    materials,_=r.h.archive_materials()
    for name,data in materials.items():(root/name).write_bytes(data)
    source_hashes={name:v.sha(ROOT/'scripts'/name) for name in r.SOURCES}
    for name in r.SOURCES:(root/'source'/name).write_bytes((ROOT/'scripts'/name).read_bytes())
    a=json.loads(r.APPROVAL.read_text());(root/'approval.json').write_bytes(r.APPROVAL.read_bytes());(root/'approved-design.md').write_bytes((ROOT/a['design']).read_bytes())
    (root/'qualification/synthetic.log').write_text('SYNTHETIC TEST ONLY; NOT TOOL QUALIFICATION\n')
    q=dict(passed=True,sourceDigests=source_hashes,designDigest=a['designDigest'],logs={'synthetic.log':v.sha(root/'qualification/synthetic.log')})
    put(root/'qualification/qualification.json',q)
    subprocess.run(['node',str(ROOT/'scripts/derive-causal-position-driver.mjs'),str(root/'worker.mjs'),str(root/'position.mjs'),str(root/'derivation.json')],check=True,capture_output=True)
    checked=subprocess.run(['node',str(ROOT/'scripts/check-causal-position-source.mjs'),str(root/'worker.mjs'),str(root/'position.mjs')],check=True,capture_output=True,text=True)
    put(root/'source-check.json',json.loads(checked.stdout))
    put(root/'runtime-materials.json',{name:v.sha(root/name) for name in ['position.mjs','worker.mjs','worker-copy.mjs','P2-inputs.json']})
    bits=[0]*40;cells=r.schedule(bits)
    reservation=dict(status='reserved',approvalDigest=v.sha(root/'approval.json'),executionRoot=str(root),runtime=dict(node='v24.18.0',v8='13.6.233.17-node.50',platform='darwin'),executable='/synthetic/node',sourceDigests=source_hashes,qualificationDigest=v.sha(root/'qualification/qualification.json'),entropyHex='00'*5,orderBits=bits,schedule=cells,implicitNodeOptions={},wakeAtStart='synthetic-no-sleep',startMonotonic=0)
    put(root/'reservation.json',reservation);rows=[];pairs=[];panels=[]
    for c in cells[:count]:
        position=c['position'];job=root/f"{position:02d}-{c['panel']}-{c['pair']:02d}-{c['orientation']}";job.mkdir();pid=100+position
        put(job/'config.json',r.config_for(job,c));(job/'entry.mjs').write_text(r.entry_source(job,c))
        (job/'entry-events.jsonl').write_text(''.join(json.dumps(dict(module=m,pid=pid))+'\n' for m in ('N','M0','M1')))
        put(job/'worker.json',dict(pid=pid,node='v24.18.0',control=c['kind']=='control',row=r.ROW,recipe={**r.RECIPE,'orders':r.ORDERS[c['orientation']]},timeOrigin=1000+position))
        put(job/'preflight.json',dict(passed=True));put(job/'preflight-first.json',dict(module=0,pid=pid));put(job/'preflight-order.json',dict(modules=[0,1],pid=pid))
        put(job/'recipe-after.json',dict(pid=pid,recipe=r.RECIPE));put(job/'completion.json',dict(completed=True,samples=2400))
        (job/'stdout.log').write_text('PRESET_PERFORMANCE_ROW_DONE cold-P2-summary 2400\n');(job/'stderr.log').write_text('')
        record=dict(**c,argv=['/synthetic/node',*r.FLAGS,str(job/'entry.mjs')],cwd=str(job),startMonotonic=position+1,startWall=position+1001,wakeBefore='synthetic-no-sleep')
        put(job/'dispatch.json',record);put(job/'exit.json',{**record,'pid':pid,'exitCode':0,'endMonotonic':position+1.5,'endWall':position+1001.5,'wakeAfter':'synthetic-no-sleep','childError':None,'wakeError':None})
        raw=samples(c,z_cost);(job/'samples.jsonl').write_text(''.join(json.dumps(x)+'\n' for x in raw));rows.append(dict(**c,**r.summarize(raw,c['orientation'])))
        if len(rows)%2==0:pairs.append(r.paired(rows[-2:]))
        if len(rows)%40==0:panels.append(r.panel_summary(pairs[-20:],c['panel']))
    passed=len(panels)==2 and all(p['passed'] for p in panels)
    put(root/'result.json',dict(status='method-qualified' if passed else 'method-not-qualified',rows=rows,pairs=pairs,panels=panels,stopReason=None,attempted=list(range(count)),notRun=list(range(count,80)),samples=count*2400,elapsedSeconds=count+2,methodQualification=passed,consumerPerformanceQualification=False,eventAssociation='unknown: retained native clocks uncalibrated'))
    reindex(root);return cells

class Tests(unittest.TestCase):
    def test_arithmetic_and_negative_coordinates(self):
        global NEGATIVES
        for bits in ([0]*40,[1]*40,[i%2 for i in range(40)]):self.assertEqual(r.schedule(bits),v.planned(bits))
        for bad in ([0]*39,[True]*40,[2]*40):
            for fn in (r.schedule,v.planned):
                with self.assertRaises(ValueError):fn(bad)
                NEGATIVES+=1
        for o in ('U','V'):
            c=r.schedule([0]*40)[0 if o=='U' else 1];raw=samples(c)
            self.assertEqual(r.summarize(raw,o),v.raw_summary(raw,o))
            for key,value in [('batch',True),('arm','plain'),('index',1),('phase','measured'),('segment','wrong'),('ms',-1),('ms',1),('constructionMs',1),('preparationMs',1),('start',float('nan')),('end',-1)]:
                bad=copy.deepcopy(raw);bad[0][key]=value
                for fn in (r.summarize,v.raw_summary):
                    with self.assertRaises((ValueError,KeyError)):fn(bad,o)
                    NEGATIVES+=1
        rows=[]
        for c in r.schedule([0]*40)[:2]:rows.append(dict(**c,**r.summarize(samples(c),c['orientation'])))
        # Independent first position 1.3, second .8 at each batch; max must not average.
        for row in rows:
            o=row['orientation'];row['p95']={'candidate':[],'reference':[]}
            for b in range(3):
                for s,a in enumerate(r.ORDERS[o][b]):row['p95'][a].append((130 if s==0 else 80) if a=='candidate' else 100)
        a=r.paired(rows);self.assertEqual(a,v.pair_result(rows));self.assertEqual(a['g'],{'first':1.3,'second':.8});self.assertEqual(a['T'],1.3)
        bad=copy.deepcopy(rows);bad[1]['pair']=99
        for fn in (r.paired,v.pair_result):
            with self.assertRaises(ValueError):fn(bad)
            NEGATIVES+=1
        for fn in (r.interval,v.band):self.assertEqual(fn(list(range(1,17))),dict(lower=2,point=8.5,upper=15))
        # Differing reference denominators distinguish median ratios (.3) from ratio medians (2).
        for b in range(3):
            for row in rows:
                o=row['orientation']
                for position,arm in enumerate(r.ORDERS[o][b]):
                    row['p95'][arm][b]=([1,20,30][b] if arm=='candidate' else [10,1,100][b]) if position==0 else ([200,30,4][b] if arm=='candidate' else [100,10,1][b])
        for fn in (r.paired,v.pair_result):
            result=fn(rows);self.assertEqual(result['g'],dict(first=.3,second=3));self.assertEqual(result['T'],3)
        for values in ([1]*15,[1]*15+[0],[1]*15+[float('inf')],[True]*16):
            for fn in (r.interval,v.band):
                with self.assertRaises(ValueError):fn(values)
                NEGATIVES+=1
    def test_exact_control_and_positive_boundaries(self):
        def make(panel,value):
            return [dict(panel=panel,pair=i,category='control' if i<2 or i>=18 else 'main',firstOrientation='U',g=dict(first=1/1.05 if i<2 or i>=18 else value,second=1.05 if i<2 or i>=18 else value),T=1.05 if i<2 or i>=18 else value) for i in range(20)]
        for fn in (r.panel_summary,v.panel_result):
            for edge in (1/1.05,1.05):self.assertTrue(fn(make('Z',edge),'Z')['passed'])
            self.assertFalse(fn(make('M',1.2),'M')['signalPassed'])
            self.assertTrue(fn(make('M',math.nextafter(1.2,math.inf)),'M')['passed'])
            for edge in (math.nextafter(1/1.05,0),math.nextafter(1.05,math.inf)):
                p=make('Z',1);p[0]['g']['first']=edge;self.assertFalse(fn(p,'Z')['controlStable'])

    def test_full_synthetic_qualification_and_reindexed_mutants(self):
        global NEGATIVES
        with tempfile.TemporaryDirectory() as tmp:
            root=Path(tmp)/'run';fixture(root)
            self.assertEqual(v.verify(root,ARCHIVE)['status'],'method-qualified')
            job=root/'00-Z-00-U'
            other=root/'01-Z-00-V';saved={}
            for name in ('worker.json','exit.json','preflight-first.json','preflight-order.json','recipe-after.json'):
                path=other/name;saved[path]=path.read_bytes();item=json.loads(saved[path]);item['pid']=100
                if name=='worker.json':item['timeOrigin']=1000
                put(path,item)
            path=other/'entry-events.jsonl';saved[path]=path.read_bytes();path.write_text(''.join(json.dumps(dict(module=m,pid=100))+'\n' for m in ('N','M0','M1')))
            reindex(root)
            with self.assertRaisesRegex(ValueError,'fresh process identity'):v.verify(root,ARCHIVE)
            NEGATIVES+=1
            for path,content in saved.items():path.write_bytes(content)
            mutations=[(root/'result.json',lambda x:x.update(status='invalid',stopReason='invented',methodQualification=False)),(root/'result.json',lambda x:x.update(methodQualification=False)),(job/'config.json',lambda x:x.update(kind='mutation')),(job/'worker.json',lambda x:x.update(timeOrigin=-1,pid=500)),(job/'preflight-order.json',lambda x:x.update(modules=[1,0])),(job/'recipe-after.json',lambda x:x['recipe'].update(warmup=99))]
            for p,change in mutations:
                before=p.read_bytes();data=json.loads(before);change(data);put(p,data);reindex(root)
                with self.assertRaises((ValueError,KeyError,OSError)):v.verify(root,ARCHIVE)
                NEGATIVES+=1;p.write_bytes(before)
            reindex(root)
            result=json.loads((root/'result.json').read_text());result['pairs'][2]['g']['first']=99;put(root/'result.json',result);reindex(root)
            with self.assertRaises(ValueError):v.verify(root,ARCHIVE)
            NEGATIVES+=1
    def test_z_failure_stops_m_and_rejects_hidden_dispatch(self):
        global NEGATIVES
        with tempfile.TemporaryDirectory() as tmp:
            root=Path(tmp)/'run';fixture(root,40,110)
            result=v.verify(root,ARCHIVE);self.assertEqual(result['status'],'method-not-qualified');self.assertEqual(result['samples'],96000)
            hidden=root/'40-M-00-U';hidden.mkdir();put(hidden/'exit.json',dict(exitCode=0));reindex(root)
            with self.assertRaises(ValueError):v.verify(root,ARCHIVE)
            NEGATIVES+=1
    def test_zero_exit_failed_validator_retained_and_reproduced(self):
        global NEGATIVES
        with tempfile.TemporaryDirectory() as tmp:
            root=Path(tmp)/'run';fixture(root,1)
            p=root/'00-Z-00-U/samples.jsonl';lines=p.read_text().splitlines();x=json.loads(lines[0]);x['constructionMs']=1;lines[0]=json.dumps(x);p.write_text('\n'.join(lines)+'\n')
            result=json.loads((root/'result.json').read_text());result.update(status='invalid',rows=[],pairs=[],panels=[],samples=0,stopReason='ValueError: cold clocks',methodQualification=False);put(root/'result.json',result)
            put(root/'failure.json',dict(stage='validate-child',position=0,verifiedRows=0,error=result['stopReason']));reindex(root)
            self.assertTrue(v.verify(root,ARCHIVE)['evidenceValid'])
            x['constructionMs']=0;lines[0]=json.dumps(x);p.write_text('\n'.join(lines)+'\n');reindex(root)
            with self.assertRaises(ValueError):v.verify(root,ARCHIVE)
            NEGATIVES+=1
    def test_preparation_failure_single_use_and_hidden_execution(self):
        global NEGATIVES
        with tempfile.TemporaryDirectory(dir=ROOT/'archive/evals') as tmp:
            root=Path(tmp).resolve()/'run'
            approval=Path(tmp)/'approval.json';grant=json.loads(r.APPROVAL.read_text());grant['output']=str(root.relative_to(ROOT));put(approval,grant)
            with patch.object(r,'OUTPUT',root),patch.object(r,'APPROVAL',approval),patch.object(r.h,'archive_materials',side_effect=ValueError('fixture material failure')):
                self.assertEqual(r.run(),1)
                with self.assertRaises(ValueError):r.run()
            self.assertTrue(v.verify(root,ARCHIVE)['evidenceValid'])
            hidden=root/'hidden';hidden.mkdir();put(hidden/'samples.jsonl',[]);reindex(root)
            with self.assertRaises(ValueError):v.verify(root,ARCHIVE)
            NEGATIVES+=2
    def test_reserved_preparation_and_zero_exit_wake_failure(self):
        with tempfile.TemporaryDirectory() as tmp:
            root=Path(tmp)/'run';fixture(root,1)
            result=json.loads((root/'result.json').read_text());result.update(status='invalid',rows=[],pairs=[],panels=[],samples=0,stopReason='ValueError: wake observation',methodQualification=False)
            put(root/'result.json',result);put(root/'failure.json',dict(stage='post-exit',position=0,verifiedRows=0,error=result['stopReason']))
            ex=root/'00-Z-00-U/exit.json';value=json.loads(ex.read_text());value.update(wakeAfter=None,wakeError='OSError: synthetic wake failure');put(ex,value);reindex(root)
            self.assertTrue(v.verify(root,ARCHIVE)['evidenceValid'])
            shutil.rmtree(root/'00-Z-00-U');(root/'runtime-materials.json').unlink();(root/'position.mjs').unlink()
            result.update(attempted=[],notRun=list(range(80)),stopReason='OSError: synthetic derivation failure');put(root/'result.json',result)
            put(root/'failure.json',dict(stage='derivation',position=None,verifiedRows=0,error=result['stopReason']));reindex(root)
            proof=v.verify(root,ARCHIVE);self.assertTrue(proof['evidenceValid']);self.assertIn('zero-dispatch',proof['failureProof'])

    def test_timeout_and_wake_failure_preserve_both(self):
        class Child:
            pid=123;returncode=None
            def wait(self,timeout=None):
                if timeout is not None:raise subprocess.TimeoutExpired('fixture',timeout)
                self.returncode=-9
            def poll(self):return self.returncode
            def kill(self):self.returncode=-9
        with tempfile.TemporaryDirectory() as tmp:
            job=Path(tmp)
            with patch.object(r.subprocess,'Popen',return_value=Child()),patch.object(r,'wake',side_effect=OSError('fixture wake')):
                result=r.execute_child(['fixture'],job,{},dict(startMonotonic=0,startWall=0),1)
            self.assertEqual(result['pid'],123);self.assertEqual(result['exitCode'],-9)
            self.assertIn('TimeoutExpired',result['childError']);self.assertIn('fixture wake',result['wakeError']);self.assertTrue((job/'exit.json').exists())

if __name__=='__main__':
    suite=unittest.defaultTestLoader.loadTestsFromTestCase(Tests)
    result=unittest.TextTestRunner(verbosity=2).run(suite)
    print(json.dumps(dict(passed=result.wasSuccessful(),tests=result.testsRun,negativeCases=NEGATIVES,syntheticOnly=True,consumerPerformanceSamples=0)))
    raise SystemExit(0 if result.wasSuccessful() else 1)
